### PR TITLE
GH-15677 Fix Match function bug

### PR DESCRIPTION
**Pull request by maurever**
Issue: #15677 

TODO:
- [x] Do not add new parameter indexes, fix the bug in our implementation according to what R::match does. Using the iris dataset and calling match with c(“setosa”, “versicolor”) should generate a new column with value 1 assigned to rows with “setosa” and 2 assigned to rows with “versicolor”). The answer should be the same with all clients. This implies that R and Python should generate the same column values.
- [x]  The default for `nomatch` should be NaN for both R and Python. In addition, only allow NaN or other numerical values for `nomatch` values. No string is supported.
- [x]  Fix the documentation to reflect the changes: `nomatch` is defaulted to NaN and can only be numerical values.  Remove the last sentences for incomparables.
- [x] Add parameter `start_index` with default value = 1, if a user wants to change indexing from 0 for example
- [x] Fix %in% to works as in base library but in different PR #15929
- [x]  Add better example in doc

### CODE REVIEW

**Review by wendycwong — CHANGES_REQUESTED** [review]

File: h2o-r/h2o-package/R/frame.R [hunk 1]
```diff
@@ -880,7 +879,7 @@ match.H2OFrame <- h2o.match
 #' @rdname h2o.match
 #' @export
 `%in%` <- function(x,table) {
-  if( is.H2OFrame(x) ) h2o.match(x,table,nomatch=0)
+  if( is.H2OFrame(x) ) h2o.match(x,table,nomatch=0) > 0
```
Comment (sebhrusen) [inline h2o-r/h2o-package/R/frame.R:882]: did you verify that the following works?
```R
> hiris <- as.h2o(iris)
> 'setosa' %in% hiris$Species
```
should it be considered as a separate bug given that `%in%` uses `match` internally?
it should work the same way as
```R
> 'setosa' %in% iris$Species
[1] TRUE
```

**Review by sebhrusen — COMMENTED** [review]
Looks good @maurever.
Added some suggestions for doc + concern about sorted numerical values.

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 2]
```diff
@@ -40,23 +38,41 @@ public String str() {
...
+    
     if (asts[2] instanceof AstNumList) {
-      matchTask = new NumMatchTask(((AstNumList) asts[2]).sort().expand(), noMatch);
+      matchTask = new NumMatchTask(((AstNumList) asts[2]).sort().expand(), noMatch, startIndex);
     }  else if (asts[2] instanceof AstNum) {
-      matchTask = new NumMatchTask(new double[]{asts[2].exec(env).getNum()}, noMatch);
+      matchTask = new NumMatchTask(new double[]{asts[2].exec(env).getNum()}, noMatch, startIndex);
     } else if (asts[2] instanceof AstStrList) {
       String[] values = ((AstStrList) asts[2])._strs;
-      Arrays.sort(values);
```
[the remark this answers is not in the document]
Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:55]: I see, the values used to be sorted because we're doing a binary search below. But then it's raising an issue: binary search requires values being sorted, but then the returned positions can be wrong…
As we're now caching the positions in a hashmap, I'd suggest to get rid of both sorting and the binary search entirely. @maurever, wdyt?

[the remark this answers is not in the document]
Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:55]: We should not sort numerical values, too. I think it is a bug. I will check and fix the numerical case as well.

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 3]
```diff
@@ -23,13 +21,13 @@
 public class AstMatch extends AstPrimitive {
   @Override
   public String[] args() {
-    return new String[]{"ary", "table", "nomatch", "incomparables"};
```
Comment (maurever) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: I try to find how incomparables parameter works in the base library. Besides the base documentation I found this thread where people struggled with the parameter. It seems it is not working as the documentation explains. 

https://stat.ethz.ch/R-manual/R-devel/library/base/html/match.html

IMO this feature is useless. So I decided not to make a PR for it.  WDYT?

**Comment by maurever** [comment]
@hannah-tillman Could you please check the documentation of this feature, please? Thanks!

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 4]
```diff
@@ -103,26 +148,53 @@ public void map(Chunk c, NewChunk nc) {
...
     }
   }
 
-  private static double in(String[] matches, String s, double nomatch) {
-    return Arrays.binarySearch(matches, s) >= 0 ? 1 : nomatch;
+  private static double in(IcedHashMap<String, Integer> matchesIndexes, String[] matches, String s, double noMatch, int startIndex) {
+    Integer mapResult = matchesIndexes.get(s);
+    int match;
+    if (mapResult == null){
+      match = Arrays.binarySearch(matches, s);
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:175]: you removed the `sort` but you keep `binarySearch`: the latter works only if `matches` is sorted, and we don't want that, so I think you should switch to `indexOf`.
A corner case performance-wise would be something like
```python
incr = [str(i) for i in range(100)]
fr = H2OFrame(incr)
fr.match(incr)  # should return a col frame equivalent to incr
decr = incr.reverse()
fr.match(decr) # should return a col frame equivalent to decr
```

**Review by hannah-tillman — APPROVED** [review]
made a few docs updates -- good on my end :)

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 4 above)
Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:175]: You are correct—my mistake. We can use sort and binarySearch, but we must save mapping to the original order of match values. Or we can use indexOf, but for `double `values, we use the `binarySearchDoublesUlp` method, so indexOf won't work correctly for doubles.

Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:175]: > We can use sort and binarySearch, but we must save mapping to the original order of match values

yeah, generating this mapping will work: for a while I thought it would be too costly to generate this, but I don't see why it would anymore, it's done in one single traversal…
We can then keep both the sorting and the binary search everywhere then, as soon as we had this mapping. Thanks @maurever

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 3 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: this param was actually unused by our implementation anyway, right? so better remove it then, thanks.

Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: weird that R failed implementing it correctly though, doesn't look that complicated (maybe it's fixed in more recent versions, didn't check)

Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: yes it is :)

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 5]
```diff
@@ -65,64 +100,124 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-      _values = values;
+    int _startIndex;
+    IcedHashMap<String, Integer> _mapping;
+    IcedHashMap<String, Integer> _matchesIndexes;
+    
+ 
+    CatMatchTask(String[] values, double noMatch, int startIndex) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        _mapping.put(values[i], i);
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: I would not override the mapping entry if it is already there: we want to return only the first index

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 6]
```diff
@@ -65,64 +100,124 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-      _values = values;
+    int _startIndex;
+    IcedHashMap<String, Integer> _mapping;
+    IcedHashMap<String, Integer> _matchesIndexes;
+
+    StrMatchTask(String[] values, double noMatch, int indexes) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        _mapping.put(values[i], i);
+      }
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:113]: see below

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 7]
```diff
@@ -65,64 +100,124 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-      _values = values;
+    int _startIndex;
+    IcedHashMap<Double, Integer> _mapping;
+    IcedHashMap<Double, Integer> _matchesIndexes;
+
+    NumMatchTask(double[] values, double noMatch, int startIndex) {
+      _mapping = new IcedHashMap<>();
+      for (int i = 0; i < values.length; i++) {
+        _mapping.put(values[i], i);
+      }
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:176]: same as above

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 5 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: we need a good test to capture that use-case

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 3 above)
Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: In the h2o.match function, we introduced this parameter, but was not working anymore and no one complained about it. :woman_shrugging:

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 5 above)
Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: Hmm, I need help understanding. The mapping is initialized only at the beginning of the MR task and used during the search.

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 3 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:26]: sure, I still don't mind removing it: better no param than not working param.

Just noticing that it's been fixed on R side since.

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java (hunk 5 above)
Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: @maurever the thing is that nothing enforces the `values` array to be made of unique values (or I missed it).
Ffor example if you refer to the R "illuminating example"
```R
## Illuminating example about NA matching
r <- c(1, NA, NaN)
zN <- c(complex(real = NA , imaginary =  r ), complex(real =  r , imaginary = NA ),
        complex(real =  r , imaginary = NaN), complex(real = NaN, imaginary =  r ))
zM <- cbind(Re=Re(zN), Im=Im(zN), match = match(zN, zN))
rownames(zM) <- format(zN)
zM ##--> many "NA's" (= 1) and the four non-NA's (3 different ones, at 7,9,10)

#--->
          Re  Im match
      NA  NA   1     1
      NA  NA  NA     1
      NA  NA NaN     1
      NA   1  NA     1
      NA  NA  NA     1
      NA NaN  NA     1
  1+NaNi   1 NaN     7
      NA  NA NaN     1
NaN+NaNi NaN NaN     9
NaN+  1i NaN   1    10
      NA NaN  NA     1
NaN+NaNi NaN NaN     9

```
we use `match(zN, zN)` and all the `NA`s match to the first occurence, not the later ones also present in `zN`.

In your implementation `values` is an array made of all `zN` values, so there are multiple `NA` there, and therefore when we loop through the values when we build the mapping, we're only interested in the first index of each occurence, typically:
```java
for (int i = 0; i < values.length; i++) {
        if (!mapping.contains(values[i])) _mapping.put(values[i], i);
}
```
do I make sense?

Comment (sebhrusen) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: so by "we need a good test to capture that use-case", I mean we need a test where the values are not all unique

Comment (maurever) [reply h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:144]: Perfect point @sebhrusen! Thanks for the explanation.

File: h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java [hunk 8]
```diff
@@ -65,64 +100,127 @@ public ValFrame apply(Env env, Env.StackHelp stk, AstRoot asts[]) {
...
-        double x = c.isNA(r) ? _noMatch : in(_values, c.atd(r), _noMatch);
+        double x = c.isNA(r) ? _noMatch : in(_matchesIndexes, _sortedValues, _mapping, c.atd(r), _noMatch, _startIndex);
         nc.addNum(x);
       }
     }
   }
 
-  private static double in(String[] matches, String s, double nomatch) {
-    return Arrays.binarySearch(matches, s) >= 0 ? 1 : nomatch;
+  private static double in(IcedHashMap<String, Integer> matchesIndexes, String[] sortedMatches, IcedHashMap<String, Integer> mapping, String s, double noMatch, int startIndex) {
```
Comment (sebhrusen) [inline h2o-core/src/main/java/water/rapids/ast/prims/search/AstMatch.java:197]: nitpick: better use `Map` interface in signatures: it's also shorter/less noisy :)

File: h2o-py/h2o/frame.py [hunk 9]
```diff
@@ -4643,26 +4643,26 @@ def stratified_split(self, test_frac=0.2, seed=-1):
...
-        >>> matchFrame = iris["C5"].match(['Iris-versicolor'])
-        >>> matchFrame
-        >>> matchFrame = iris["C5"].match(['Iris-setosa'])
-        >>> matchFrame
+        >>> match_col = iris["C5"].match(['Iris-setosa', 'Iris-versicolor'])
+        >>> match_col.names = ['match']
+        >>> iris_match = iris.cbind(match_col)
+        >>> iris_split = iris_match.split_frame(ratios=[0.05], seed=1)[0]
+        >>> print(iris_split)
         """
```
Comment (sebhrusen) [inline h2o-py/h2o/frame.py:4664]: is it possible to update this example to show the result as mentioned in previous comment https://github.com/h2oai/h2o-3/pull/15896#discussion_r1395969684?
It clarifies a lot what the function is doing, esp. for online doc.

same with R

File: h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_match.py [hunk 10]
```diff
@@ -4,28 +4,72 @@
...
-    assert_is_type(matchFrame, H2OFrame)    # check return type
-    assert matchFrame.sum()[0,0]==150.0, "h2o.H2OFrame.match() command is not working."  # check return result
+    match_frame = iris["C5"].match(['Iris-setosa', 'Iris-versicolor'], nomatch=nomatch)
+    assert_is_type(match_frame, H2OFrame, nomatch=0)    # check return type
+    assert round(match_frame.sum()[0, 0]) == 150, "h2o.H2OFrame.match() command is not working."  # check return result
+    assert match_frame[0, 0] == 1, "match value should be 1"
+    assert match_frame[50, 0] == 2, "match value should be 2"
+    assert match_frame[100, 0] == nomatch, "match value should be 0"
 
+    match_values = ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
```
Comment (sebhrusen) [inline h2o-py/tests/testdir_apis/Data_Manipulation/pyunit_h2oH2OFrame_match.py:32]: can you please add a test with duplicate values to be sure?
```match_values = match_values + match_values``` ?
result should be the same if I'm correct

**Review by sebhrusen — COMMENTED** [review]
looking good, thanks @maurever 
Apparently, you still need to rebase to resolve conflicts, so I will approve next time.

**Comment by maurever** [comment]
Thanks, @sebhrusen; the PR is updated based on your suggestion and rebased.

**Review by wendycwong — APPROVED** [review]

File: h2o-py/h2o/frame.py [hunk 11]
```diff
@@ -4653,26 +4653,29 @@ def stratified_split(self, test_frac=0.2, seed=-1):
...
 
         :examples:
 
-        >>> iris = h2o.import_file("http://h2o-public-test-data.s3.amazonaws.com/smalldata/iris/iris.csv")
-        >>> matchFrame = iris["C5"].match(['Iris-versicolor'])
-        >>> matchFrame
-        >>> matchFrame = iris["C5"].match(['Iris-setosa'])
-        >>> matchFrame
+        >>> data = h2o.import_file("h2o://iris")
+        >>> match_col = data["C5"].match(['Iris-setosa', 'Iris-versicolor'])
```
Comment (sebhrusen) [inline h2o-py/h2o/frame.py:4672]:
```suggestion
        >>> match_col = data["class"].match(['Iris-setosa', 'Iris-versicolor'])
```

**Review by sebhrusen — COMMENTED** [review]
tested the Py doc generation with the result rendered in the example, it works nicely!
Also updated the suggestion to use real data from the example.

Maybe same could be done in R.

**Comment by maurever** [comment]
@hannah-tillman, please check that the documentation is generated correctly. @sebhrusen also suggested adding output into the R example, but I think it will not work in the R. If you have any idea how to improve the R example to be similar to the Python example, please let me know (or feel free to commit this change here). Thanks!

File: h2o-py/h2o/frame.py (hunk 11 above)
Comment (sebhrusen) [reply h2o-py/h2o/frame.py:4672]: example fails if using "C5" as column name: `h2o://iris` has named columns, contrary to the previous csv file.

**Review by sebhrusen — DISMISSED** [review]
LGTM, thanks @maurever 

I took the initiative to commit a small fix in the Py example.
Regarding the R example, looking at R doc, I don't think it's possible to render the results the same way as we can in Py, common practice seems to be using a comment, like:
```R
sample ##--> describe shortly what is expected when this is printed
```
for example, in this case
```R
sample ##--> `match` column should be made of `1` for `setosa`, `2` for `versicolor` and `NaN` for `virginica`
```

**Comment by hannah-tillman** [comment]
@maurever  The python docs build fine 👍 

For the R documentation, I went through the current R guide to find an example of something adding the output because I didn’t really remember that happening. What I was able to find was a rare case of [output being added as a comment](https://docs.h2o.ai/h2o/latest-stable/h2o-r/docs/reference/h2o.relevel.html#examples), so we could do the same for this? Would look like this:
```
#' \dontrun{
#' h2o.init()
#' data <- as.h2o(iris)
#' match_col <- h2o.match(data$Species, c("setosa", "versicolor", "setosa"))
#' iris_match <- h2o.cbind(data, match_col)
#' sample <- h2o.splitFrame(iris_match, ratios=0.05, seed=1)[1]
#' sample
#' # [[1]]
#' #   Sepal.Length Sepal.Width Petal.Length Petal.Width    Species C1
#' # 1          5.2         3.5          1.5         0.2     setosa  1
#' # 2          5.0         3.5          1.3         0.3     setosa  1
#' # 3          7.0         3.2          4.7         1.4 versicolor  1
#' # 4          4.9         2.4          3.3         1.0 versicolor  1
#' # 5          5.5         2.4          3.8         1.1 versicolor  1
#' # 6          5.8         2.7          5.1         1.9  virginica  0
#' #
#' # [12 rows x 6 columns] 
#' }
```
This is the only example I was able to find for output in the R docs, though, so it is currently the only idea i have. Let me know if this is what you want.

**Comment by maurever** [comment]
Thanks, @seb and @hannah-tillman. The PR is now ready for review.

**Review by sebhrusen — APPROVED** [review]
LG, thank you @maurever !
